ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -0,0 +1,523 @@
+   * </p>
+   * <p>
+   *   There is a slight chance that an {@link Account} could be updated successfully, but its value was
+   *   set based on an outdated {@link Account}. This can be fixed after {@code generationId} is introduced
+   *   to {@link Account}.
+   * </p>
+   */
+  @Override
+  public boolean updateAccounts(Collection<Account> accounts) {
+    checkOpen();
+    if (accounts == null) {
+      throw new IllegalArgumentException("accounts cannot be null");
+    }
+    long startTimeMs = System.currentTimeMillis();
+    logger.trace("Start updating to HelixPropertyStore with accounts={}", accounts);
+    if (hasDuplicateAccountIdOrName(accounts)) {
+      return false;
+    }
+    boolean hasSucceeded = false;
+    // make a pre check for conflict between the accounts to update and the accounts in the local cache. Will fail this
+    // update operation for all the accounts if any conflict exists. There is a slight chance that the account to update
+    // conflicts with the accounts in the local cache, but does not conflict with those in the helixPropertyStore. This
+    // will happen if some accounts are updated but the local cache is not refreshed.
+    if (!hasConflictWithCache(accounts)) {
+      hasSucceeded = helixStore.update(FULL_ACCOUNT_METADATA_PATH, currentData -> {
+        ZNRecord newRecord;
+        if (currentData == null) {
+          logger.debug(
+              "ZNRecord does not exist on path={} in HelixPropertyStore when updating accounts. Creating a new ZNRecord.",
+              FULL_ACCOUNT_METADATA_PATH);
+          newRecord = new ZNRecord(String.valueOf(ZN_RECORD_ID));
+        } else {
+          newRecord = currentData;
+        }
+        Map<String, String> accountMap = newRecord.getMapField(ACCOUNT_METADATA_MAP_KEY);
+        if (accountMap == null) {
+          logger.debug("AccountMap does not exist in ZNRecord when updating accounts. Creating a new accountMap");
+          accountMap = new HashMap<>();
+        }
+        AccountInfoMap remoteAccountInfoMap = new AccountInfoMap(accountMap);
+
+        // if there is any conflict with the existing record, fail the update. Exception thrown in this updater will
+        // be caught by Helix and eventually helixStore#update will return false.
+        if (hasConflictingAccount(accounts, remoteAccountInfoMap)) {
+          String message =
+              "Updating accounts failed because one or more accounts to update conflict with accounts in helix property store";
+          // Do not depend on Helix to log, so log the error message here.
+          logger.error(message);
+          throw new IllegalArgumentException(message);
+        } else {
+          for (Account account : accounts) {
+            try {
+              accountMap.put(String.valueOf(account.getId()), account.toJson().toString());
+            } catch (Exception e) {
+              String message = new StringBuilder(
```

Comment:
you don't need to use a `StringBuilder` here, the compiler will do that if you just use string concatenation (which is easier to read).